SLING-5768 Introduce rep:slingResourceTypes

ghenzler opened the pull request:
Introduce rep:slingResourceTypes as extension to Oak permission system

ghenzler committed:
SLING-5768 Introduce rep:slingResourceTypes as extension to Oak
permission system

ghenzler committed:
SLING-5768 improvements

* moved classes to impl package
* allow for resourceType@path values

rombert commented on contrib/extensions/sling-oak-restrictions/pom.xml:
```diff
@@ -0,0 +1,90 @@
+				<configuration>
+					<signature>
+						<groupId>org.codehaus.mojo.signature</groupId>
+						<artifactId>java16</artifactId>
```
This should not be needed if you inherit from the sling parent pom.

rombert commented on contrib/extensions/sling-oak-restrictions/pom.xml:
```diff
@@ -0,0 +1,90 @@
+		<dependency>
+			<groupId>org.apache.jackrabbit</groupId>
+			<artifactId>oak-core</artifactId>
+			<version>1.2.7</version>
```
please make the scope provided, since you're importing transitive dependencies in the classpath, which makes it hard to see the actual dependencies.

You're using guava in the classes but they're not explicitly imported.

rombert commented on contrib/extensions/sling-oak-restrictions/src/main/java/org/apache/sling/slingoakrestrictions/impl/ResourceTypePattern.java:
```diff
@@ -0,0 +1,139 @@
+    @Override
+    public boolean matches(@Nonnull Tree tree, @Nullable PropertyState property) {
+        String actualResourceType = TreeUtil.getString(tree, SLING_RESOURCE_TYPE);
+        boolean isResourceTypeMatch = resourceTypes.contains(actualResourceType);
```
Have you considered allowing for resource type inheritance?

rombert commented on contrib/extensions/sling-oak-restrictions/src/main/java/org/apache/sling/slingoakrestrictions/impl/SlingRestrictionProviderImpl.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Override
+    public void validateRestrictions(String oakPath, @Nonnull Tree aceTree) throws AccessControlException {
```
is this needed?

ghenzler replied:
It's calling the base class at org.apache.jackrabbit.oak.spi.security.authorization.restriction.AbstractRestrictionProvider.validateRestrictions(String, Tree) which performs sanity checks - I think it makes sense.

rombert replied:
My point was that you're overriding the method just to call the superclass method. Don't you achieve the same effect my simply removing the method?

ghenzler replied:
You are right (the oak class was explicitly overriding because there was additional checks). Will remove it.

rombert commented on contrib/extensions/sling-oak-restrictions/src/main/java/org/apache/sling/slingoakrestrictions/impl/ResourceTypePattern.java:
```diff
@@ -0,0 +1,139 @@
+        List<String> plainResourceTypes = new ArrayList<String>();
+        Map<String,Set<String>> resourceTypesByPath = new HashMap<String,Set<String>>();
+        for (String resourceTypeRaw : resourceTypesRaw) {
+            if(resourceTypeRaw.contains("@")) {
```
Nitpick - if you're using Guava you might want to use `Splitter` as well.

ghenzler committed:
SLING-5768 Improvements (e.g. second restriction type
rep:resourceTypesWithChildren)

- renamed rep:slingResourceTypes to rep:resourceTypes for brevity (a
name clash here is not likely)
- Introduced rep:resourceTypesWithChildren to allow matching nodes with
a resource type and all its children
- cleaned up pom (to use inherited elements where possible)
- removed dependency to guava 
- added junit tests

ghenzler committed:
SLING-5768 removed unnecessary override of validateRestrictions and
reduced some log levels to trace

ghenzler committed:
SLING-5768 Changing from rep:* to sling:* namespaces for restrictions

ghenzler committed:
SLING-5768 Also changing constant names for rep:* to sling:* namespace
change
